src/vcpkg/versions.cpp

```diff
@@ -50,93 +51,185 @@ namespace vcpkg::Versions
+    static const char* parse_skip_number(const char* const s, uint64_t* const n)
     {
-        std::regex relaxed_scheme_match("^(0|[1-9][0-9]*)(\\.(0|[1-9][0-9]*))*");
-
-        if (!std::regex_match(str, relaxed_scheme_match))
+        const char ch = *s;
+        if (ch == '0')
         {
-            return Strings::format(
-                "Error: String `%s` must only contain dot-separated numeric values without leading zeroes.", str);
+            *n = 0;
+            return s + 1;
         }
-
-        return RelaxedVersion{str, Util::fmap(Strings::split(str, '.'), [](auto&& strval) -> uint64_t {
-                                  return as_numeric(strval).value_or_exit(VCPKG_LINE_INFO);
-                              })};
+        if (ch < '1' || ch > '9')
+        {
+            return nullptr;
+        }
+        size_t i = 1;
+        for (; Parse::ParserBase::is_ascii_digit(s[i]); ++i)
+        {
+        }
+        *n = as_numeric({s, i}).value_or_exit(VCPKG_LINE_INFO);
+        return s + i;
     }
 
-    ExpectedS<SemanticVersion> SemanticVersion::from_string(const std::string& str)
+    ExpectedS<RelaxedVersion> RelaxedVersion::from_string(const std::string& str)
     {
-        // Suggested regex by semver.org
-        std::regex semver_scheme_match("^(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)"
-                                       "(?:-((?:0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9][0-9]*|[0-9]"
-                                       "*[a-zA-Z-][0-9a-zA-Z-]*))*))?"
-                                       "(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$");
+        // (0|[1-9][0-9]*)(\.(0|[1-9][0-9]*))*;
 
-        if (!std::regex_match(str, semver_scheme_match))
+        const char* cur = str.c_str();
+        std::vector<uint64_t> numbers;
+        for (;;)
         {
-            return Strings::format(
-                "Error: String `%s` is not a valid Semantic Version string, consult https://semver.org", str);
+            numbers.push_back(0);
+            cur = parse_skip_number(cur, &numbers.back());
+            if (!cur || (*cur != 0 && *cur != '.'))
+            {
+                return Strings::format(
+                    "Error: String `%s` must only contain dot-separated numeric values without leading zeroes.", str);
+            }
+            if (*cur == 0)
+            {
+                break;
+            }
+            // *cur == '.'
+            ++cur;
         }
 
-        SemanticVersion ret;
-        ret.original_string = str;
-        ret.version_string = str;
+        return RelaxedVersion{str, std::move(numbers)};
+    }
 
-        auto build_found = ret.version_string.find('+');
-        if (build_found != std::string::npos)
+    // 0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*
+    // equivalent:
+    static const char* skip_prerelease_identifier(const char* const s)
+    {
+        auto cur = s;
+        for (; Parse::ParserBase::is_ascii_digit(*cur); ++cur)
```

Comment:
```suggestion
        while (Parse::ParserBase::is_ascii_digit(*cur)) { ++cur; }
```

Reply:
It's funny, I implemented all of these as for loops because I _thought_ I remembered you saying you preferred
`for (;;) {}` over `while (1) {}`.